kratos/includes/registry_item.h

```diff
@@ -228,11 +292,39 @@ class KRATOS_API(KRATOS_CORE) RegistryItem
 
     RegistryItem& GetItem(std::string const& rItemName);
 
-    template<typename TDataType> TDataType const& GetValue() const
+    template <typename TDataType>
+    void EvaluateValue()
     {
         KRATOS_TRY
 
-        return *(std::any_cast<std::shared_ptr<TDataType>>(mpValue));
+        using TFunctionReturnType = std::shared_ptr<TDataType>;
+        using TFunctionType = std::function<TFunctionReturnType()>;
+
+        // Assign callable value
+        TFunctionType func = std::any_cast<TFunctionType>(mpValue);
+        mpCallable = func();
+
+        // Set value name
+        std::stringstream buffer;
+        buffer << *(std::any_cast<TFunctionReturnType>(mpCallable));
+        mValueName = buffer.str();
+
+        KRATOS_CATCH("");
+    }
+
+    template <typename TDataType>
+    TDataType const &GetValue()
+    {
+        KRATOS_TRY
+
+        // This is executed the first time we access the GetValue for this item
+        using TFunctionReturnType = std::shared_ptr<TDataType>;
+
+        if (std::any_cast<TFunctionReturnType>(&mpCallable) == nullptr) {
+            this->EvaluateValue<TDataType>();
+        }
```

Comment:
[`std::optional`](https://en.cppreference.com/w/cpp/utility/optional)

Reply:
I've never used this `std::optional`. From the standard documentation I guess you mean using
```c++
if (!std::optional(&mpCallable)) {
    this->EvaluateValue<TDataType>();
}
```
Am I wrong?

Reply:
Not sure how is the best way to apply it in this case. `EvaluateValue` do more things aside from assigning the return of the lambda (mainly initialize the value of the variable so the Registry can be serialized)

If we were to apply an `std::optional` here that would mean that we have to either:
1 - Split the function in `EvalVariable` and `EvalVariableValue` and only use the `optional`in the first half (because we don't want the code that assigns the name to be executed more than once)
2 - Make the `EvaluateVariable` return something like `std::optional<std::reference_wrapper<std::pair<TFunctionRetyrnType,std::string>>>` . Also I am not sure if this prevents the code of  `EvaluateVariable` from being executed or not.

Not an expert on the use of `std::optional`. If I am misusing this glad to see how its done :)

Reply:
`std::optional` is a wrapper for objects that may or may not have a value.

For example, if you default construct a smart pointer, it's value will be in an invalid state (`nullptr`) until you replace it with something it can point to. Unfortunately, someone can still dereference the pointer while it's in an invalid state, leading to a segfault. `std::optional` stores a flag indicating whether the objects has a valid value, and will throw an exception if someone tries to access it in its invalid state. The point is that it effectively forces people to think about what happens if the object doesn't have a value, and provides a default behaviour (throw an exception) when they forget to do so.

Example
```cpp
struct S
{
    std::optional<std::unique_ptr<int>> mpWhatever;
};

int main()
{
    S s; // <== default constructor, mpWhatever has no value
    std::cout << *s.mpWhatever.value(); // <== exception, mpWhatever has no value

    s.mpWhatever = std::make_unique(1); // <== s.mpWhatever now has a valid value
    std::cout << *s.mpWhatever.value(); // <== 1
}
```

So, I suggest wrapping `mpCallable` with `std::optional` whenever it's assigned to.

```cpp
mpCallable = std::make_optional(func());
```

Then the check would become uglier, but we wouldn't run into segfaults if we forgot to check the pointer elsewhere.
```cpp
using TOptionalFunctionReturnType = std::optional<TFunctionReturnType>;
if (!std::any_cast<TOptionalFunctionReturnType>(&mpCallable).has_value()) {
    this->EvaluateValue<TDataType>();
}
```

Reply:
Oh!, ok :+1: